File: tests/testthat/test-step_order.R
# The guard on the hand-written step order, R side.
#
# The pipeline's step order is spelled out by hand in five scripts, three of
# them R: R/siab_main.R, R/run_testdata.R and tests/fixtures/make_r_dumps.R.
# Nothing generates them from one definition, and the user ruled on 2026-09-21c
# that the duplication stays (contention C7). What replaces a generator is
# tests/step_order.json and these tests: the file states the order once, each
# script declares the steps it leaves out and the arguments it passes
# differently, and a change made in some scripts and not the others fails here.
#
# The failure this is for is the silent one. A step added to two files of
# three, or an argument changed in one, leaves every comparison running and no
# longer testing what it claims to test. The comparisons cannot see it: they
# check values, and a pipeline that skipped a step still produces values.
#
# tests/pytest/test_step_order.py is the same guard over the two Python
# scripts, reading the same file.

step_order <- jsonlite::fromJSON(here::here("tests", "step_order.json"),
                                 simplifyVector = FALSE)

canonical <- unlist(step_order$steps)

r_scripts <- Filter(function(spec) spec$arm == "R", step_order$scripts)

# The sequence of step calls a script makes, in the order it makes them.
# Comment lines are dropped first, which is what keeps the commented-out calls
# a script declares as skipped from counting as calls, and what keeps the
# project's long explanatory comments from matching on a step name they discuss.
step_calls <- function(file) {
  lines <- readLines(here::here(file), warn = FALSE)
  lines <- lines[!grepl("^\\s*#", lines)]
  calls <- character(0)
  for (line in lines) {
    hits <- regmatches(line, gregexpr(
      paste0("\\b(", paste(canonical, collapse = "|"), ")\\s*\\("), line))[[1]]
    calls <- c(calls, sub("\\s*\\($", "", hits))
  }
  calls
}

# Every value a script passes as `handling =`, which is the one argument the
# runners and the dump writers deliberately disagree on.
handling_arguments <- function(file) {
  lines <- readLines(here::here(file), warn = FALSE)
  lines <- lines[!grepl("^\\s*#", lines)]
  hits <- regmatches(lines, gregexpr("handling\\s*=\\s*\"[a-z]+\"", lines))
  unique(gsub(".*\"([a-z]+)\".*", "\\1", unlist(hits)))
}

test_that("tests/step_order.json lists each step once", {
  expect_equal(anyDuplicated(canonical), 0L)
  expect_true(length(canonical) > 0)
})

for (file in names(r_scripts)) {
  spec <- r_scripts[[file]]
  skipped <- names(spec$skips)

  test_that(paste(file, "runs the steps in the canonical order"), {
    expect_true(all(skipped %in% canonical),
                info = "a skip names a step the definition does not list")
    expect_equal(step_calls(file), setdiff(canonical, skipped))
  })

  test_that(paste(file, "passes the arguments it declares"), {
    expect_equal(handling_arguments(file), spec$arguments$handling)
  })
}

File: tests/pytest/test_step_order.py
"""
The guard on the hand-written step order, Python side.

The pipeline's step order is spelled out by hand in five scripts, two of them
Python: python/main.py and tests/pytest/make_py_dumps.py. Nothing generates
them from one definition, and the user ruled on 2026-09-21c that the
duplication stays (contention C7). What replaces a generator is
tests/step_order.json and these tests: the file states the order once, each
script declares the steps it leaves out and the arguments it passes
differently, and a change made in some scripts and not the others fails here.

The failure this is for is the silent one. A step added to one arm and not the
other, or an argument changed in one script, leaves every comparison running
and no longer testing what it claims to test. The comparisons cannot see it:
they check values, and a pipeline that skipped a step still produces values.

tests/testthat/test-step_order.R is the same guard over the three R scripts,
reading the same file.
"""

from __future__ import annotations

import json
import re

import pytest

from conftest import PROJECT

DEFINITION = json.loads((PROJECT / "tests" / "step_order.json").read_text())
CANONICAL: list[str] = DEFINITION["steps"]
SCRIPTS = {name: spec for name, spec in DEFINITION["scripts"].items()
           if spec["arm"] == "python"}


def _uncommented(path: str) -> list[str]:
    """The script's lines with the comment-only ones dropped.

    Both scripts declare a skipped step by commenting its call out, and both
    carry long explanatory comments that name steps they do not call, so a
    parser that reads comments would see calls that do not happen.
    """
    text = (PROJECT / path).read_text().splitlines()
    return [line for line in text if not line.lstrip().startswith("#")]


def step_calls(path: str) -> list[str]:
    """The sequence of step calls a script makes, in the order it makes them.

    Both scripts run every step through a local `run()` helper, which takes the
    step as its first argument, after the store in the dump writer. Names that
    are not steps, the helpers' own parameters among them, are dropped.
    """
    pattern = re.compile(r"\brun\(\s*(?:con\s*,\s*)?([A-Za-z_][A-Za-z_0-9]*)")
    called = pattern.findall("\n".join(_uncommented(path)))
    return [name for name in called if name in CANONICAL]


def handling_arguments(path: str) -> list[str]:
    """Every value a script passes as `handling=`.

    That is the one argument the runners and the dump writers deliberately
    disagree on: the dumps sort parallel episodes on tenure, because the wage
    setting sorts on draws that differ between the arms by construction.
    """
    hits = re.findall(r"handling\s*=\s*\"([a-z]+)\"", "\n".join(_uncommented(path)))
    return sorted(set(hits))


def test_the_definition_lists_each_step_once():
    assert len(set(CANONICAL)) == len(CANONICAL)
    assert CANONICAL


@pytest.mark.parametrize("path", sorted(SCRIPTS))
def test_the_script_runs_the_steps_in_the_canonical_order(path):
    skipped = set(SCRIPTS[path]["skips"])
    assert skipped <= set(CANONICAL), "a skip names a step the definition does not list"
    assert step_calls(path) == [s for s in CANONICAL if s not in skipped]


@pytest.mark.parametrize("path", sorted(SCRIPTS))
def test_the_script_passes_the_arguments_it_declares(path):
    assert handling_arguments(path) == [SCRIPTS[path]["arguments"]["handling"]]
